### ISSUE TITLE
Zeno x Gentopia -- Eval Integration

### ISSUE THREAD

**Issue by billxbf** [zeno-ml/instance-views#9]
## Context
Hi @cabreraalex @neubig,  I'm Bill leading [Gentopia](https://github.com/Gentopia-AI/Gentopia), a framework to build LLM-Driven Agents with extensive and config-driven interface.  Now there's a pool to register agents and call each other in hierarchical orders.  As discussed with Professor @neubig on discord, it will be helpful to create a new `view` extending `openai-chat` to reflect intermediate tool-calls and custom agent evaluation.

## Significace 
- This would allow Zeno to show tool-augmented LMs' reasoning traces, which is important if one wants to tune some agents.
- This would allow Zeno to show more flexible eval because sometimes there's no `Ground Truth` or `Expected` answers.
- This completes Gentopia since it misses visualization of agent evaluation.

## Todo 
Basically our evaluation data needs a new view called `tool-chat` (or some better name) with two changes to be made from `openai-chat`.

- A "tool" role (or "function" role as openai named it)  and a little icon (like :hammer_and_wrench: ) so it can render intermediate tool response.
- Remove "Expected" and make the section fully customizable to input (can it render markdown?). because the eval tasks not only contain  QAs but also some custom eval instructions (eg. unit test cases for coding)

![image](https://github.com/zeno-ml/instance-views/assets/65674752/0c78aa39-e0bc-4eed-9b3f-da3aa2b849c2)

**Comment by cabreraalex** [comment]
This is awesome thanks so much for the context and idea!

The changes make lots of sense. Some clarifying questions:
* Is the format for the data the same as the ChatGPT API? e.g. a JSON with roles and content?
* Does the "tool" role always return text?
* Will "expected" always be markdown? 

It may help if you have one or multiple example files with the data format. I can then play around and get a version to you this week!

**Comment by billxbf** [comment]
Hi Alex, thanks for your help! Here are the things you need.  

- The format is the same as ChatGPT Function Call [API](https://platform.openai.com/docs/guides/gpt/function-calling), namely a JSON with role and content. The role field can be one of `system`, `user`, `assistant`, `function`
- Yes the `tool` (or `function`, whichever you like)  will always be string type.
- Yes, the "expected" will always be markdown.


Here's an eval example to refer

(*Above the Bar*)
```
[ 
{"role": "system", "content": "dr_science is an expert in math and natural science. He is able to answer complex questions about math, physics, chemistry, biology, and more."},
{"role": "user", "content": "what is 2322*2311"},
{"role": "function", "name": "wolfram_alpha", "content": "5366142"},
{"role": "assistant", "content": "The product of 2322 and 2311 is 5366142."}
]
```
(*Below the Bar -- render markdown input, default to empty*)
```
Grader: GatedGrader
Status: Passed
EM: 1.0
xxxxxx (some other custom eval info)
```

And I'm open to change at your convenience! Thanks again~

**Comment by cabreraalex** [comment]
This looks perfect! I'll give it a go in the next day or two and get back to you!

**Comment by neubig** [comment]
I also kinda wonder: maybe we could solve this by just expanding the `openai-chat` interface by adding markdown support to all cells. Once we do that, we could then just preface each tool response with `**Tool**\n`, which I think should be obvious enough.

**Comment by billxbf** [comment]
Yea that should give more flexibility to the message pad in the future.
For example, ReWOO has multiple tool calls in a single round (check below) and will look cleaner if the `**Tool**` field supports markdown.

```
{"role": "function", "name": "ReWOO Planner", 
"content": "
Plan : Search for more information about Melanie C.
#E1 = Wikipedia[Melanie C]
Plan : Find out who featured on Melanie C’s number one single Never Be The Same Again in
2000.
#E2 = LLM[Who featured on Melanie C’s number one single Never Be The Same Again in
2000? Given context: #E1 ]
Plan : Search for more information about the featured artist.
#E3 = Wikipedia[ #E2 ]
Plan : Find out if the featured artist died in a car accident in 2002.
#E4 = LLM[Did #E2 die in a car accident in 2002? Given context: #E2 ]
"},
```

Tell me if I can help with anything!

**Comment by cabreraalex** [comment]
Quick update I'm playing around with this, hoping to wrap it up tomorrow or Tuesday!

**Comment by billxbf** [comment]
No rush at all!

**Comment by cabreraalex** [comment]
Alright I have a first version! All it does currently is render the assistant block and label from markdown. I'm happy to make it render the other blocks (user, system) as markdown as well if that's a common use case. Here's a screenshot and the minimal code to test it out.

<img width="646" alt="Screenshot 2023-08-30 at 20 11 54" src="https://github.com/zeno-ml/instance-views/assets/4563691/f0d788b3-e8c7-4e91-882e-759ae754b87d">


```python
import zeno
from zeno import ZenoParameters
import pandas as pd


df = pd.DataFrame({
    "data": [
        [
            {"role": "assistant", "content": "# Hello \n * first bullet \n * second bullet"},
            {"role": "user", "content": "# Hello \n * first bullet \n * second bullet"},
         ],
    ]
})

zeno.zeno(ZenoParameters(metadata=df, data_column="data", view="openai-chat-markdown"))
```

**Comment by billxbf** [comment]
Looks Great! My only problem is if the `Expected` label here can be optional or customizable 

https://github.com/zeno-ml/instance-views/blob/2229d262760b395aeefbabe129b9f1ea1ba8b702/openai-chat-markdown/src/InstanceView.svelte#L63

Because sometimes eval doesn't have an "expected" ground truth. (eg. Detecting Toxicity in the conversation)

**Comment by cabreraalex** [comment]
For sure, if you don't pass in a labelColumn option to Zeno then it won't show the expected part :)

**Comment by billxbf** [comment]
Yeah, is it possible to pass in a labelColumn but not showing the "Expected" Title? 
(So a pure markdown box below)

**Comment by cabreraalex** [comment]
OH I see, yeah of course let me update it :)

**Comment by cabreraalex** [comment]
I'll make it never show "expected" for this markdown view, expect the user to add context to the label themselves.

**Comment by cabreraalex** [comment]
Updated it! If you have any example data would love to run it through and see

**Comment by billxbf** [comment]
Thanks a lot, that looks perfect ~ I'll need to make some changes in Gentopia to use the new view. Let me send you the final effect afterward!

**Comment by neubig** [comment]
Hey @billxbf , just wanted to go full-circle on this. Would there be a way we could be helpful here?

**Comment by billxbf** [comment]
Hi Graham, sorry I was refactoring some code to use this new view, but didn't finish it last weekend. 

I'll try to make it work and update you by EOW.

**Comment by neubig** [comment]
Great!

**Comment by billxbf** [comment]
Hi @neubig and @cabreraalex, sorry for the late update but I've refactored something in GentPool, so that the default view for eval should look like below. It contains an Agent conversation + a Grader scoring (if configured so).

`$ python evaluate.py openai_template`

![image](https://github.com/zeno-ml/instance-views/assets/65674752/c2160a27-f33c-41dd-896c-1698ec339acb)

LGTM! I'll make a release announcement for this on the weekend highlighting this visualization. I'm also considering making a video tutorial covering it. (I'm moving to the West Coast soon so things are kind of piled up on my side, but will do when I find time.)

Anyway, great thanks for your work! Tell me if you need anything as well :)

**Comment by cabreraalex** [comment]
AMAZING! it looks super cool. Are you also measuring/quantifying the "grade" results in Zeno? Happy to help set that up too to filter through the data and find interesting instances/model behaviors

**Comment by billxbf** [comment]
Thanks Alex! I think this would be good for now because I don't see a need for batch measuring of agents. At most times I run an agent on 10 tasks, look at what's wrong, and modify its component accordingly. I'll open a separate issue if needed :) 

Now I find it promising to use tool-calling agents to generate instruction-tuning datasets and to fine-tune the LLM itself with the augmented responses. This can work around the painful latency and cost issue of agents while getting targeted distribution shift and model output. To this point, I'll probably switch back to the original `openai-chat` view.

**Comment by cabreraalex** [comment]
Perfect, sounds good! Let me know if I can help out with anything

**Comment by billxbf** [comment]
Looking forward to your talk! Glad to see this progress :)
